### PR TITLE
Setup Kotlin, Ktlint and Dokka; convert all Xtend classes to Kotlin

### PR DESCRIPTION
**Pull request by Vuksaa**

### CODE REVIEW

File: alchemist/alchemist-implementationbase/src/main/kotlin/it/unibo/alchemist/kotlin/AlchemistExtensions.kt [hunk 1]
```diff
@@ -0,0 +1,4 @@
+package it.unibo.alchemist.kotlin
+
+fun <E> E.unfold(extractor: (E) -> Sequence<E>): Sequence<E> =
```
Comment (DanySK) [inline alchemist/alchemist-implementationbase/src/main/kotlin/it/unibo/alchemist/kotlin/AlchemistExtensions.kt]: Missing Kdoc

File: alchemist/alchemist-implementationbase/src/main/kotlin/it/unibo/alchemist/model/implementations/layers/BidimensionalGaussianLayer.kt [hunk 2]
```diff
@@ -0,0 +1,19 @@
+package it.unibo.alchemist.model.implementations.layers
+
+import it.unibo.alchemist.model.implementations.utils.BidimensionalGaussian
+import it.unibo.alchemist.model.interfaces.Layer
+import it.unibo.alchemist.model.interfaces.Position
+
+open class BidimensionalGaussianLayer @JvmOverloads constructor(
```
Comment (DanySK) [inline alchemist/alchemist-implementationbase/src/main/kotlin/it/unibo/alchemist/model/implementations/layers/BidimensionalGaussianLayer.kt]: Missing Kdoc

File: alchemist/alchemist-implementationbase/src/main/kotlin/it/unibo/alchemist/model/implementations/neighborhoods/SimpleNeighborhood.kt [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+import it.unibo.alchemist.model.interfaces.Environment
+import it.unibo.alchemist.model.interfaces.Neighborhood
+import it.unibo.alchemist.model.interfaces.Node
+import org.danilopianini.util.ArrayListSet
+import org.danilopianini.util.Hashes
+import org.danilopianini.util.ListBackedSet
+import org.danilopianini.util.ListSet
+import org.danilopianini.util.ListSets
+
+class SimpleNeighborhood<T> private constructor(
```
Comment (DanySK) [inline alchemist/alchemist-implementationbase/src/main/kotlin/it/unibo/alchemist/model/implementations/neighborhoods/SimpleNeighborhood.kt]: Missing Kdoc

File: alchemist/alchemist-implementationbase/src/main/kotlin/it/unibo/alchemist/model/implementations/neighborhoods/Neighborhoods.kt [hunk 4]
```diff
@@ -0,0 +1,14 @@
+package it.unibo.alchemist.model.implementations.neighborhoods
+
+import it.unibo.alchemist.model.interfaces.Environment
+import it.unibo.alchemist.model.interfaces.Node
+
+class Neighborhoods {
```
Comment (DanySK) [inline alchemist/alchemist-implementationbase/src/main/kotlin/it/unibo/alchemist/model/implementations/neighborhoods/Neighborhoods.kt]: Missing Kdoc

File: alchemist/alchemist-implementationbase/src/main/kotlin/it/unibo/alchemist/model/implementations/terminators/StableForSteps.kt [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+import com.google.common.collect.Maps
+import com.google.common.collect.Table
+import com.google.common.collect.Tables
+import it.unibo.alchemist.model.interfaces.Environment
+import it.unibo.alchemist.model.interfaces.Molecule
+import it.unibo.alchemist.model.interfaces.Node
+import it.unibo.alchemist.model.interfaces.Position
+import java.util.function.Predicate
+
+open class StableForSteps<T>(
```
Comment (DanySK) [inline alchemist/alchemist-implementationbase/src/main/kotlin/it/unibo/alchemist/model/implementations/terminators/StableForSteps.kt]: close this class, Missing Kdoc

File: alchemist/alchemist-implementationbase/src/main/kotlin/it/unibo/alchemist/model/implementations/terminators/StableForSteps.kt [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    private val checkInterval: Long,
+    private val equalIntervals: Long
+) : Predicate<Environment<T>> {
+    var success: Long = 0
```
Comment (DanySK) [inline alchemist/alchemist-implementationbase/src/main/kotlin/it/unibo/alchemist/model/implementations/terminators/StableForSteps.kt]: private

File: alchemist/alchemist-implementationbase/src/main/kotlin/it/unibo/alchemist/model/implementations/terminators/StableForSteps.kt [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    var positions: Map<Node<T>, Position> = emptyMap()
```
Comment (DanySK) [inline alchemist/alchemist-implementationbase/src/main/kotlin/it/unibo/alchemist/model/implementations/terminators/StableForSteps.kt]: private

File: alchemist/alchemist-implementationbase/src/main/kotlin/it/unibo/alchemist/model/implementations/terminators/StableForSteps.kt [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+    var contents = makeTable<T>(0)
```
Comment (DanySK) [inline alchemist/alchemist-implementationbase/src/main/kotlin/it/unibo/alchemist/model/implementations/terminators/StableForSteps.kt]: private

File: alchemist/alchemist-implementationbase/src/main/kotlin/it/unibo/alchemist/model/implementations/utils/BidimensionalGaussian.kt [hunk 9]
```diff
@@ -0,0 +1,24 @@
+package it.unibo.alchemist.model.implementations.utils
+
+import org.apache.commons.math3.analysis.BivariateFunction
+import org.apache.commons.math3.util.FastMath.exp
+import java.io.Serializable
+
+open class BidimensionalGaussian(
```
Comment (DanySK) [inline alchemist/alchemist-implementationbase/src/main/kotlin/it/unibo/alchemist/model/implementations/utils/BidimensionalGaussian.kt]: Missing Kdoc, if class can be closed, close it

File: alchemist/alchemist-implementationbase/src/test/kotlin/it/unibo/alchemist/test/TestNeighborhood.kt [hunk 10]
```diff
@@ -0,0 +1,22 @@
+package it.unibo.alchemist.test
+
+import it.unibo.alchemist.model.implementations.environments.Continuous2DEnvironment
+import it.unibo.alchemist.model.implementations.neighborhoods.Neighborhoods
+import it.unibo.alchemist.model.implementations.nodes.IntNode
+import org.junit.Assert
+import org.junit.Test
+
+class TestNeighborhood {
```
Comment (DanySK) [inline alchemist/alchemist-implementationbase/src/test/kotlin/it/unibo/alchemist/test/TestNeighborhood.kt]: Missing Kdoc

File: alchemist/alchemist-swingui/src/main/kotlin/it/unibo/alchemist/boundary/gui/view/properties/RangedDoubleProperty.kt [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+import javafx.beans.property.DoublePropertyBase
+import java.io.ObjectInputStream
+import java.io.ObjectOutputStream
+import java.io.Serializable
+import java.lang.reflect.Type
+
+private const val LOWER_BOUND = "lower bound"
+private const val UPPER_BOUND = "upper bound"
+
+open class RangedDoubleProperty @JvmOverloads constructor(
```
Comment (DanySK) [inline alchemist/alchemist-swingui/src/main/kotlin/it/unibo/alchemist/boundary/gui/view/properties/RangedDoubleProperty.kt]: Missing Kdoc, can this one be closed?

File: alchemist/build.gradle [hunk 12]
```diff
@@ -5,9 +5,12 @@ buildscript {
         maven { url "https://plugins.gradle.org/m2/" }
     }
     dependencies {
+        classpath "gradle.plugin.org.jmailen.gradle:kotlinter-gradle:$ktlintVersion"
         classpath "org.danilopianini:build-commons:${buildCommonsVersion}"
         classpath "org.danilopianini:javadoc.io-linker:${javadocIOLinkerVersion}"
-        classpath "org.xtext:xtext-gradle-plugin:${xtextPluginVersion}"
+//        classpath "org.xtext:xtext-gradle-plugin:${xtextPluginVersion}"
```
Comment (DanySK) [inline alchemist/build.gradle]: delete commented line

File: alchemist/alchemist-swingui/src/main/kotlin/it/unibo/alchemist/boundary/gui/view/properties/RangedDoubleProperty.kt (hunk 11 above)
Comment (DanySK) [reply alchemist/alchemist-swingui/src/main/kotlin/it/unibo/alchemist/boundary/gui/view/properties/RangedDoubleProperty.kt]: Not checking for null in equals is always a problem. You're welcome to fix.

Comment (Vuksaa) [reply alchemist/alchemist-swingui/src/main/kotlin/it/unibo/alchemist/boundary/gui/view/properties/RangedDoubleProperty.kt]: I just noticed that the first method call is a safe call through the ?. operator, my bad.

**Comment by DanySK** [comment]
Looks good to go

**Pull request merged**